newsrevenuehub/rev-engine
# DEV-2980: Revise amount editor

Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

#### What's this PR do?

- Adds an ElementEditor component which is a frame for element editors that uses the new hooks created in DEV-2765.
- Revises the amount editor to reflect new designs.

#### Why are we doing this? How does it help us?

- Improves code consistency.
- Better UX.

#### How should this be manually tested? Please include detailed step-by-step instructions.

To test basic adding of amounts:

1. Edit a page.
2. Edit the Contribution Amount element.
3. Enter a number and click the plus button for each frequency.
4. Confirm that the amounts stay correctly ordered (e.g. if you add an amount of 100 to one that has 50 and 200, the 100 appears in the middle, not at the end).
5. Confirm changes save successfully.
6. Confirm you're able to make a contribution to the page and the updated amounts appear in the correct order.

To test field validation:
1. Edit a page's Contribution Amount element.
2. Try to add a negative number and confirm it's not allowed.
3. Try to add a number with too many decimal places, like 1.234, and confirm it's not allowed.
4. Try to add a word like `no`, You shouldn't able to type it or paste it into the field.
5. Try to add an amount that has already been added.

To test setting the default amount:
1. Edit a page's Contribution Amount element.
2. Click new amounts to change the default for each frequency.
3. Save changes and confirm the defaults are reflected on the published page.

To test removing an amount:
1. Edit a page's Contribution Amount element.
2. Remove an amount from each frequency.
3. Save changes and confirm the amounts are updated on the published page.
4. Confirm you can't remove the last amount for a frequency.

To test the "allow other" checkbox:
1. Edit a page's Contribution Amount element.
5. Change the "allow other" checkbox's value and save changes.
7. Confirm the changes are reflected on the published page.
8. Repeat steps 1-3 but changing the value to the opposite one you had before.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

Screenshots may need updating.

#### Have automated unit tests been added? If not, why?

Yes.

#### How should this change be communicated to end users?

n/a

#### Are there any smells or added technical debt to note?

No.

#### Has this been documented? If so, where?

No.

#### What are the relevant tickets? Add a link to any relevant ones.

https://news-revenue-hub.atlassian.net/browse/DEV-2980

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No.

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

No.
